### PR TITLE
Fix recompilation issue with content store

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113533

While running the accuracy minifier, I was getting the error:
```
NotImplementedError("xor_sum only implemented with inductor")
```

The logs showed that the cache limit was exceeded, and it was falling back to
eager mode which doesn't work for this function. The cache failures was due to
the code guarding on the id of the function being compiled which in this case is
a closure that gets re-created for each function call so the guard always fails.

This fixes the issue by making the storage hash kernel a global function and
working around the dynamo dependency by the `lazy_compile` helper which defers
the `torch.compile` call to the first invocation.

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/utils/_content_store.py [hunk 1]
```diff
@@ -43,6 +44,52 @@
...
+    def decorate_fn(fn):
+        @functools.wraps(fn)
+        def compile_hook(*args, **kwargs):
+            compiled_fn = torch.compile(fn, **compile_kwargs)
+            globals()[fn.__name__] = functools.wraps(fn)(compiled_fn)
+            return compiled_fn(*args, **kwargs)
+
+        return compile_hook
+
+    return decorate_fn
```
Comment (ezyang) [inline torch/utils/_content_store.py:62]: Thanks, we probably want to move this somewhere else eventually, maybe another utils module.

File: test/test_content_store.py [hunk 2]
```diff
@@ -60,6 +60,13 @@ def test_scalar(self, device):
         # Should not raise an error
         hash_storage(torch.tensor(2, device=device).untyped_storage())
 
+    @torch._dynamo.config.patch(cache_size_limit=1)
+    def test_repeated_hash(self, device):
+        # Test that repeated hashing doesn't trigger a recompile in dynamo
+        # If it does, we will execute prims.xor_sum in eager which fails
+        for _ in range(4):
+            hash_storage(torch.tensor(2, device=device).untyped_storage())
```
Comment (ezyang) [inline test/test_content_store.py:68]: FWIW, recreating the same closure in a loop shouldn't actually cause recompile if the code object stays the same and there are no free vars. cc @Fidget-Spinner who had a patch to fix this but I can't remember if it actually landed

Comment (peterbell10) [reply test/test_content_store.py:68]: This test was failing for me on a fairly recent main branch commit.

Comment (Fidget-Spinner) [reply test/test_content_store.py:68]: I just re-landed that patch in main two days ago https://github.com/pytorch/pytorch/pull/109427

Comment (peterbell10) [reply test/test_content_store.py:68]: I just checked and cherry-picking that commit does make the test pass, oh well.

Comment (ezyang) [reply test/test_content_store.py:68]: It's fine, lazy compile at top level is a useful thing to have anyway
